To-do 테스트 작성하기

yxxnhx opened the pull request:
# 목표
To-do 테스트 작성하기

## 진행사항

### 2-3일차
강의를 듣고 List 파일을 할일이 있을 경우, 없을경우를 나누어 테스트 파일을 생성하였습니다
현재 jest, tdd에 대해서 많은 지식이 없어서 다른 분들은 어떻게 하셨는지 참고하여서 describe와 context에 대해 공부하면서 작성하고 있습니다.

### 4일차
트레이너님 조언을 참고해서 List test 파일 수정 및 Input test를 생성하였습니다

### 5일차
트레이너님 조언을 참고해서 List와 Input에서 render까지 외부로 함수로 빼서 사용하였습니다
현재 App test 파일을 생성하여 랜더링 여부와 추가 버튼을 누르면 할일이 추가되는 것까지 테스트를 하였는데 완료 버튼을 누르면 삭제되고, 목록이 없으면 없다는 메시지가 보이는 것을 어떻게 테스트를 해봐야할지 고민입니다. Item과 동일하게 task 배열을 만들어서 하면 될지 조금 더 시도를 해봐야겠습니다.

### 6일차 
#### 1차
-  App test 파일을 구현하였습니다.
추가 버튼과 삭제 버튼을 누르면 입력한 값이 제대로 출력 되고 삭제되는지 여부를 확인할 수 있도록 하였습니다.
![image](https://user-images.githubusercontent.com/50559373/198813919-b9736d7b-5bd5-402a-b506-5350aa349a8f.png)
![image](https://user-images.githubusercontent.com/50559373/198813936-71430d74-5bda-4e9a-80b9-a5aadfb14d92.png)
그런데 위와 같이 계속해서 동일한 코드가 겹쳐서 그 부분을 바깥으로 빼서 정리를 하고 싶은데 it 외부에 뺄 시에는 계속 동일한 오류가 납니다.
따로 정리를 할 수 있는 방법이 없을까요?
![image](https://user-images.githubusercontent.com/50559373/198814207-792728ce-0ff9-4b08-96eb-314927cac435.png)
그리고 궁금한 것이 있는데 프로젝트 규모가 더 커져서 페이지가 여러 개가 되고, 나눈 컴포넌트가 여러가지일 경우에도 현재와 같이 동일하게 src 폴더 안에 기존 jsx파일과 test.jsx 파일을 함께 두는 것이 맞나요?
아니면 test 폴더를 따로 생성해서 test 파일만 관리하는 것이 맞나요?

#### 2차
-  불필요한 코드를 지우고 테스트 목적에 맞게 입력된 값이 보인다, 추가 버튼이 랜더링 된다는 것으로 나누어 테스트 코드를 재구성하였습니다.
저는 조금 더 명확하게 하기 위하여 버튼들을 변수화하여 사용을 했었는데 불필요한 것인가요?

### 7일차
아직 부족한 점이 많아서 사소한 것도 자꾸 놓치네요ㅠㅠ
불필요한 코드를 지우고, 테스트 목적에 맞게 테스트 이름 변경 및 자주 쓰는 테스트용 데이터는 tasks로 빼서 fixtures 폴더에 분리시켰습니다

yxxnhx committed:
To-do 테스트 작성하기

List 파일을 할일이 있을 경우, 없을경우를 나누어 테스트 파일을 생성

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+
+describe('List', () => {
+  const onClickDelete = jest.fn();
+  const renderElement = (task) => (
```
`Element` 보다 좀더 의도가 드러나는 단어를 사용하게 좋을 것 같아요!

yxxnhx replied:
네 변수명을 조금 더 고민해보겠습니다

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+import List from './List';
+
+describe('List', () => {
+  const onClickDelete = jest.fn();
```
props 명이 아니므로, handleXXX으로 하는게 더 좋을 거 같아요!

yxxnhx replied:
아하 제가 자꾸 헷갈리네요ㅠㅠ 변경하겠습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+import List from './List';
+
+describe('List', () => {
+  const onClickDelete = jest.fn();
+  const renderElement = (task) => (
```
맥락 구분을 위해 서로 띄어줍시다!ㅎㅎ

yxxnhx replied:
넵 반영하여 수정하겠습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,46 @@
+  });
+
+  context('List에서 할 일이 있을 때', () => {
+    it('입력한 할 일이 출력된다', () => {
```
이것도 "렌더링된다"라고 표현하는게 더 좋을 것 같아요!ㅎㅎ

yxxnhx replied:
넵 랜더링으로 수정하겠습니다

daadaadaah reviewed (changes requested):
빠른 PR에 더불어 테스트코드도 잘 작성하셔서 너무 기분이 좋네요!👏👏👏
다음 PR 기다리고 있을게요~ㅎㅎ

yxxnhx committed:
Input 테스트 파일 작성

입력값과 추가 버튼을 누를 시 제대로 호출이 되는지 여부 test

yxxnhx committed:
Refector: List test 컴포넌트 수정

의도가 드러날 수 있게 변수명 수정

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,38 @@
+describe('<Input />', () => {
+  const handleClick = jest.fn();
+  const handleChange = jest.fn();
+  const InputElement = (
+    <Input
+      value="잠자기"
+      onClick={handleClick}
+      onChange={handleChange}
+    />
+  );
```
render 함수도 중복되는데, render 함수 까지 같이 빼지 않은 이유가 있으신가용?ㅎㅎ

yxxnhx replied:
아하 그냥 안에 내용만 따로 빼야 된다고 생각했는데 render까지 빼도 괜찮군요! 그렇다면 render까지 함께 빼겠습니다~!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,38 @@
+  it('Input에 입력한 값이 제대로 랜더링되는가', () => {
+    render(InputElement);
+
+    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
+    fireEvent.change(input, {
+      target: {
+        value: '잠자기',
+      },
+    });
+    expect(input).toHaveAttribute('value', '잠자기');
```
맥락 구분을 위해 30줄이랑 31줄, 31줄이랑 36줄 각각 다 한줄씩 띄어주시는게 좋아요!(다른 부분도 마찬가지구요!)

yxxnhx replied:
아 넵넵 맥락 구분까지 염두해서 다음부터는 코드를 짜고 정리할 수 있도록 하겠습니다

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,38 @@
+    expect(handleClick).toBeCalled();
+  });
+
+  it('Input에 입력한 값이 제대로 랜더링되는가', () => {
```
List.test.jsx 파일에서 기재한 것(`~렌더링된다.`)처럼 표현해주시면 좋을 것 같아요!(16번 테스트도요!)

yxxnhx replied:
넵! 수정하겠습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+describe('List', () => {
+  const handleClickDelete = jest.fn();
+
+  const listElement = (task) => (
+    <List
+      tasks={task}
+      onClick={handleClickDelete}
+    />
+  );
```
https://github.com/CodeSoom/react-week3-assignment-1/pull/155#discussion_r1007538376 이 리뷰 참고해주세요!ㅎㅎ

yxxnhx replied:
넵 반영하겠습니다!

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    it('할 일이 없어요 문구 랜더링된다', () => {
+      const task = [];
+      const { getByText } = render(listElement(task));
+      getByText('할 일이 없어요!');
```
expect 문이 빠진 것 같아요!ㅎㅎ

yxxnhx replied:
앗 그렇네요 추가하겠습니다!

yxxnhx committed:
Page test 구현

Page 컴포넌트에 할일, 추가, 삭제 버튼이 랜더링되는지 여부 확인

yxxnhx committed:
review: Input test 파일 수정

render까지 외부에 빼서 코드 수정 및 test 문구 수정

yxxnhx committed:
review: List page 수정

listElement에 render까지 빼서 코드 정리

yxxnhx committed:
review: List test expect 추가

할 일이 없어요 랜더링 되는 expect 문을 추가하였습니다

yxxnhx committed:
App test 생성

랜더링 여부, 추가 버튼을 누를 시 할일이 추가되는 것까지 완료

yxxnhx committed:
App test 구현

할일이 있을 시와 없을 시를 구분하여 handleClick, handleDelete를 실행 여부 확인

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,67 @@
+      expect(getByText(/할 일이 없어요!/)).not.toBeNull();
+    });
+
+    it('값을 입력하면 "추가" 버튼이 랜더링 된다', () => {
+      const { getByLabelText, getByText, container } = appElement();
+      const input = getByLabelText('할 일');
+
+      fireEvent.change(input, {
+        target: {
+          value: taskTitle,
+        },
+      });
+
+      const addTaskButton = getByText('추가');
+
+      fireEvent.click(addTaskButton);
+
+      expect(container).toHaveTextContent(taskTitle);
+    });
```
이 테스트의 의도가 궁금합니다! `값을 입력하면, 입력된 값이 보인다` 이거나 `추가 버튼이 렌더링 된다`로 테스트가 이루어질 것 같은데, 둘이 섞인 것 같아서요!

yxxnhx replied:
앗 그렇네요 값이 보이는 것과 추가 버튼이 랜더링 되는 것 두 개로 나누어 작성하겠습니다!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,36 @@
+  const handleChangeTitle = jest.fn();
+  const handleAddTask = jest.fn();
+
+  it('Page에 할일, 추가, 삭제 버튼이 있는가', () => {
```
테스트 이름 수정해주세요!ㅎㅎ

yxxnhx replied:
앗 수정하겠습니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,67 @@
+import App from './App';
+
+describe('<App />', () => {
+  const appElement = () => render((<App />));
```
함수명에 의도가 좀더 드러나면 좋을 것 같아요! appElement 만 보면, 처음 코드를 본 동료가 어떤 함수인지 의문이 들 것 같아요!

yxxnhx replied:
넵 조금 더 의도에 맞게 수정하겠습니다

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,67 @@
+        },
+      });
+
+      const addTaskButton = getByText('추가');
```
버튼들 변수화해서 쓰시는데, 이유가 궁금합니다!ㅎㅎ

yxxnhx replied:
조금 더 명확하게 버튼을 사용하고 싶어서 했었는데 불필요한가요?
불필요하다면 변수화하지 않고 바로 사용하겠습니다!

daadaadaah replied:
때에 따라서 다르겠지만, 지금 상황에서 "추가"라는 텍스트가 있어서 변수화 하지 않아도 될 것 같아요!ㅎㅎ
혹시, 버튼이라는 걸 드러내고 싶으면, 아래처럼 `getbyRole`를 활용할 수 있을 것 같아요!ㅎㅎ
```js
getbyRole('button', { name: /추가 })
```

yxxnhx replied:
아하 그렇군요 그렇다면 getbyRole을 활용하여 다시 해보겠습니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,67 @@
+  });
+
+  context('할 일이 있을 시', () => {
+    it('"삭제" 버튼이 랜더링된다', () => {
```
테스트에 삭제 버튼이 렌더링 되는지만 확인하면 될 것 같아요! 테스트 목적에 맞게 불필요한 코드들은 지워봅시다!

yxxnhx replied:
넵 불필요한 코드는 지우겠습니다~!

daadaadaah reviewed (changes requested):
> 그런데 위와 같이 계속해서 동일한 코드가 겹쳐서 그 부분을 바깥으로 빼서 정리를 하고 싶은데 it 외부에 뺄 시에는 계속 동일한 오류가 납니다.
따로 정리를 할 수 있는 방법이 없을까요?
- 저라면 중복되는 거 함수로 뺄 것 같은데, 어떻게 빼셔서 어떤 오류가 났는지 보여주시면, 답변하는데 더 도움이 될 것 같아요!ㅎㅎ


> 그리고 궁금한 것이 있는데 프로젝트 규모가 더 커져서 페이지가 여러 개가 되고, 나눈 컴포넌트가 여러가지일 경우에도 현재와 같이 동일하게 src 폴더 안에 기존 jsx파일과 test.jsx 파일을 함께 두는 것이 맞나요?
아니면 test 폴더를 따로 생성해서 test 파일만 관리하는 것이 맞나요?
- 이건 팀마다 다를 것 같아요! 아샬님이 함께 두는 편이라, 보통 함께 둡니다!ㅎㅎ

yxxnhx committed:
refactor: App test, Page test 수정

겹치는 fireEvent를 renderApp에 넣어서 구조분해할당으로 사용
불필요한 코드 삭제
테스트 의도에 맞게 코드 재구성

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,73 @@
+
+    return {
+      ...result,
+      changeInput: (value) => {
+        const element = result.getByLabelText('할 일');
+        fireEvent.change(element, {
+          target: {
+            value,
+          },
+        });
+      },
```
App.test.jsx 모든 테스트에 ChangeInput 이라는 함수가 필요한 건가요?

yxxnhx replied:
값을 넣어 테스트하는 경우에는 모두 필요해서 renderApp에 넣어서 구조분해할당으로 사용을 했는데 그렇게 하면 안될까요?
그냥 따로 함수를 선언해서 사용하는 것이 더 좋을까요?
안 된다면 왜 안되는지 설명해주시면 감사하겠습니다!🥹

daadaadaah replied:
구조분해할당 같은 기술을 사용해보시는거 잘하셨어요!👍
다만, 지금 App.test.jsx를 보면, 값을 넣어 테스트 하는 부분도 있고 아닌 부분도 있어서 말씀드린거였어요!ㅎㅎ 만약, 모든 App.test.jsx 테스트에 changeInput이 필요한 경우라면, 있어도 되겠지만, 그게 아니라서요!
만약, 값을 넣어서 테스트 하는 부분이 중복되어 함수로 묶어 주고 싶다면, 그 부분만 묶어주셔서 사용해주셔도 됩니다!ㅎㅎ

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,73 @@
+      expect(getByText(/할 일이 없어요!/)).not.toBeNull();
+    });
+
+    context('값을 입력하면', () => {
```
`값`보다 `할일`이라고 표현해주시는게 좀더 의도가 드러나는 것 같아요!ㅎㅎ

yxxnhx replied:
넵 의도에 맞게 수정하겠습니다~!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,36 @@
+describe('<Input />', () => {
+  const handleClick = jest.fn();
+  const handleChange = jest.fn();
+  const InputElement = () => render((<Input
```
함수 이름은 함수의 역할이 드러나는 표현을 적어주시는게 좋아요! 지금 단순히 Input 요소가 아니라 Input 요소가 렌더링되는게 함수 역할이 아닐까 생각해요!(다른 테스트에도 마찬가지로 적용시켜주세요!)

yxxnhx replied:
넵 변경해보겠습니다

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,36 @@
+
+describe('<Input />', () => {
+  const handleClick = jest.fn();
+  const handleChange = jest.fn();
+  const InputElement = () => render((<Input
```
맥락 구분을 위해 한줄 띄어줍시당!ㅎㅎ

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,36 @@
+
+    expect(handleClick).not.toBeCalled();
+
+    fireEvent.click(getByText('추가'));
+    expect(handleClick).toBeCalled();
```
맥락 구분을 위해 한줄 띄어줍시당!ㅎㅎ(다른 곳도 마찬가지로 적용시켜주세요!)

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    expect(handleClick).toBeCalled();
+  });
+
+  it('Input에 입력한 값이 제대로 랜더링 된다', () => {
+    InputElement();
+
+    const input = screen.getByPlaceholderText('할 일을 입력해 주세요');
+
+    fireEvent.change(input, {
+      target: {
+        value: '잠자기',
+      },
+    });
+
+    expect(input).toHaveAttribute('value', '잠자기');
+  });
```
1. 28줄이 없어도 34줄이 통과될 거예요! '잠자기2'로 value 변경했을 때, '잠자기2'가 보이는지 한번 테스트해보세요! 만약, 테스트가 안된다면, 왜 안되는지에 대해 한번 고민해봅시다!ㅎㅎ
2. 이런 문제가 발생하지 않기 위해서 함수를 묶어줄 때 어떻게 해야될지도 함께 고민해보시는게 좋을 것 같아요!ㅎㅎ

yxxnhx replied:
지워서 테스트를 돌려보니 이미 기본 값으로 들어가 있는 상태여서 해당 부분이 필요가 없었던 것이었네요😅
그래서 함수로 묶은 부분에서 그냥 값을 넣지 않고 변수로 빼서 테스트를 할 때 그 인자로 받은 변수의 값을 넣어주는 방식으로 바꾸어보았습니다
이 방법이 맞는 것을까요? 아니라면 혹시 다른 방법이 있을까요?

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+
+  context('List에서 할 일이 있을 때', () => {
+    it('입력한 할 일이 랜더링된다', () => {
+      const tasks = [{
+        id: 1,
+        title: '잠자기',
+      },
+      {
+        id: 2,
+        title: '누워있기',
+      },
+      ];
```
tasks 가 다른 테스트에서도 사용되고 있어요! 이렇게 여러 곳에서 사용되는 데스트용 데이터를 보통 fixture라고 이야기합니다. 저는 이러한 데이터를 `fixtures` 폴더 하나를 만들어서 따로 관리하는 편이에요!

yxxnhx replied:
넵 분리하여 수정해보겠습니다~!

daadaadaah commented on src/Page.test.jsx:
```diff
@@ -0,0 +1,36 @@
+  const handleChangeTitle = jest.fn();
+  const handleAddTask = jest.fn();
+
+  it('Page에 할일, 추가, 삭제 버튼이 랜더링 된다', () => {
+    const { container, getByText } = render((
+      <Page
+        tasks={tasks}
+        taskTitle={taskTitle}
+        onChangeTitle={handleChangeTitle}
+        onClickAddTask={handleAddTask}
+        onClickDeleteTask={handleDeleteTask}
+      />));
+
+    expect(container).toHaveTextContent('완료');
+
+    fireEvent.click(getByText('추가'));
+    expect(handleAddTask).toBeCalled();
+  });
```
테스트 이름에 맞게 테스트 코드가 작성되지 않은 것 같아요! 한번 체크해주세요!

yxxnhx replied:
아하 그렇네요 테스트 명을 수정해보겠습니다

daadaadaah commented on src/List.test.jsx:
```diff
@@ -0,0 +1,47 @@
+  });
+
+  context('List에서 할 일이 있을 때', () => {
+    it('입력한 할 일이 랜더링된다', () => {
```
`입력한`은 빼고 작성해주시는게 좋을 것 같아요! 지금은 그냥 props로 받은 tasks가 렌더링 되는 것을 확인하는 테스트이지, 사용자가 입력해서 상태가 변경되는지 확인하는 테스트가 아니라고 생각하기 때문이예요!ㅎㅎ

yxxnhx replied:
그렇네요.. 테스트를 짤 때 조금 더 고민을 해봐야겠군요ㅠㅠ

yxxnhx committed:
review: App test 수정

버튼 명확히 드러내기 위해 호출 방식 변경
테스트명  의도에 맞게 변경

yxxnhx committed:
review: List test 수정

함수 이름 역할이 명시적으로 드러날 수 있게 수정
맥락 구분을 위해 코드 정리
필요 없는 fireEvent 삭제 및 value 값 변경할 수 있도록 수정

yxxnhx committed:
feat: fixtures 생성

테스트용 데이터 생성

yxxnhx committed:
review: List test 코드 정리

자주 쓰는 테스트용 데이터를 fixtures로 빼서 코드 정리

yxxnhx committed:
review: Page test 수정

테스트 목적에 맞게 테스트 이름 수정
자주 쓰는 테스트 데이터 fixtures로 정리하여 수정

yxxnhx committed:
review: List test 수정

테스트 목적에 맞게 테스트 명 수정

daadaadaah commented on src/App.test.jsx:
```diff
@@ -0,0 +1,73 @@
+    });
+
+    context('값을 입력하면', () => {
+      it('onChange가 실행된다', () => {
```
App 컴포넌트의 관심사에 대해 한번 생각해보면, onChange 실행이 아니라 상태가 변경된다고 기재하시는게 더 좋을 것 같아요!

daadaadaah commented on src/Input.test.jsx:
```diff
@@ -0,0 +1,32 @@
+  const handleClickAddTask = jest.fn();
+  const handleChangeTitle = jest.fn();
+
+  const InputElement = (taskTitle = '') => render((<Input
```
render랑 같이 빼주셨으면, 함수명도 그에 맞게 바꿔주시는게 좋을 거 같아요!ㅎㅎ
App.test.jsx에서는 하셔서 어떻게 하시는지 아시는 것 같은데, 깜박하신거 같아요!ㅎㅎ(다른 부분도 마찬가지로 반영해주세요!ㅎㅎ)

daadaadaah reviewed (changes requested):
이번주도 수고 많으셨습니다!ㅎㅎ 다음주에도 같이 힘내봐요👊
